### DISCUSSION TITLE
IHostBuilder as default builder

### DISCUSSION THREAD

**Discussion by ParadiseFallen** [AvaloniaUI/Avalonia#6157] [Ideas]
I suggest to use IHostBuilder for avalonia. then we can easy setup di and other options. use configuration files. etc.
then we dont need Splat anymore.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0

**Comment by kekekeks** [comment]
Give it a try and see if you could make it to not blow up on macOS and if you could make IHost::StartAsync to return a Task instead of blocking.

**Reply by ParadiseFallen** [reply]
Avalonia framework can be a IHostedService

**Reply by kekekeks** [reply]
Not really.
1) on desktop we have to run the event loop _synchronously_
2) on macOS we have to run the event loop synchronously on the **first** application thread (that's it, the thread where `Main` was called, and that has to be a normal `Main`, not the async one)
3) on mobile we don't have any control over startup procedure whatsoever, startup hooks are invoked by the target platform

**Reply by ParadiseFallen** [reply]
maui .net6 will use IHostBuilder for running an app on mobile devices. Probably when it released you can easy swap startup

**Reply by ParadiseFallen** [reply]
Also https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0#ihostenvironment
there Start method that run host in sync

**Reply by kekekeks** [reply]
>maui .net6 will use IHostBuilder for running an app on mobile devices

_Will use_ or _uses_? Because when looking into their code base I see this https://github.com/dotnet/maui/blob/4340ccf4f87ffae1bf146c5ffe4908e53bd72879/src/Core/src/Hosting/Internal/AppHost.cs#L26-L42

```cs
		public async Task StartAsync(CancellationToken cancellationToken = default)
		{
			_logger?.Starting();

			await Task.Run(() => { });

			_logger?.Started();
		}

		public async Task StopAsync(CancellationToken cancellationToken = default)
		{
			_logger?.Stopping();

			_ = await Task.FromResult(new object());

			_logger?.Stopped();
		}

		public void Dispose() => DisposeAsync().AsTask().GetAwaiter().GetResult();
```

I might be looking in a wrong place though.

**Reply by kekekeks** [reply]
Anyway, give it a try. Our AppBuilder is relatively small, shouldn't be that hard to replace it.

**Reply by ParadiseFallen** [reply]
It will use. maui in preview and will be released later. For now i wanna see easy di in avalonia. For now i need to use custom extension method that register deps in Splat and then resolve deps when model created. For MVVM navigation there can be helper like CreateVM<ViewModelName>(/*extra arguments*/). It can use ActivatorUtilities inside

**Reply by ParadiseFallen** [reply]
Target is also not to use Splat as DI container. Use build in microsoft di container for avalonia

**Reply by kekekeks** [reply]
Splat is used by ReactiveUI. Avalonia doesn't have any DI container preference.
